### PR TITLE
API to extract WKB from geometry value (fix #162)

### LINKED ISSUE #162: API to extract WKB from geometry value
In diffs, values of geometry columns have geometry encoded according to GPKG spec: http://www.geopackage.org/spec/#gpb_format - the actual WKB is wrapped with some extra information.

When working with diffs in clients, it would be useful to have API call that would extract WKB from the geometry value so that the clients do not need to implement the conversion themselves.

### PR DESCRIPTION
**Pull request by alexbruy**
Add new function `createWkbFromGpkgHeader()` to extract geometry in WKB format from the geometry encoded according to GPKG spec.

Fix #162.

### CODE REVIEW

**Comment by alexbruy** [comment]
Not sure where is the right place for it, so I have put it in the same file where other functions related to geometry handling are located. As I have no prior experience with exposing and using C API in Python new function is not public yet.

**Comment by PeterPetrik** [comment]
You need to add C - wrapper to https://github.com/lutraconsulting/geodiff/blob/master/geodiff/src/geodiff.h
and Python-wrapper in https://github.com/lutraconsulting/geodiff/blob/master/pygeodiff/main.py

please add tests too

**Comment by alexbruy** [comment]
Tried to add wrappers. Not sure if this is the correct way of doing it, but I have never worked with a such low-level stuff before.

**Comment by PeterPetrik** [comment]
looks good. can you add python test here? : https://github.com/lutraconsulting/geodiff/tree/master/pygeodiff/tests

File: geodiff/src/geodiff.cpp [hunk 1]
```diff
@@ -1060,3 +1060,8 @@ bool GEODIFF_CT_columnIsPkey( GEODIFF_ChangesetTableH tableHandle, int i )
 {
   return static_cast<ChangesetTable *>( tableHandle )->primaryKeys.at( i );
 }
+
+const char *GEODIFF_G_getWkb( const char *gpkgWkb )
```
Comment (PeterPetrik) [inline geodiff/src/geodiff.cpp:1064]: I would call it: GEODIFF_createWkbFromGpkgHeader ; similarly the python function. Can you please improve docs in geodiff.h to exactly specify what is does , maybe with example?

File: geodiff/tests/test_geometry_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
+/*
+ GEODIFF - MIT License
+ Copyright (C) 2020 Martin Dobias
```
Comment (PeterPetrik) [inline geodiff/tests/test_geometry_utils.cpp:3]: 2022 Alex?

File: geodiff/tests/test_geometry_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#include "geodiff_testutils.hpp"
+#include "geodiff.h"
+
+#include "changesetreader.h"
+#include "sqliteutils.h"
+
+#include <iostream>
+
+
+TEST( GeometryUtilsTest, test_wkb_from_geometry )
```
Comment (PeterPetrik) [inline geodiff/tests/test_geometry_utils.cpp:16]: we still need python test too (smoke test is ok, just to make sure the python wrapper works)

File: geodiff/src/geodiff.h [hunk 4]
```diff
@@ -502,6 +503,19 @@ GEODIFF_EXPORT void GEODIFF_V_getData( GEODIFF_ValueH valueHandle, char *data );
 GEODIFF_EXPORT void GEODIFF_V_destroy( GEODIFF_ValueH valueHandle );
 
 
+/**
+ * Extracts WKB geometry from the geometry encoded according to GeoPackage spec.
+ * Strips GeoPackageBinaryHeader from the StandardGeoPackageBinary envelope and
+ * returns geometry as WKB. More details at https://www.geopackage.org/spec/#gpb_format
+ */
```
Comment (PeterPetrik) [inline geodiff/src/geodiff.h:510]: can you add description of the args + return (like return value + length is buffer you need to free in GEODIFF_free)

File: pygeodiff/geodifflib.py [hunk 5]
```diff
@@ -435,6 +435,19 @@ def read_changeset(self, changeset):
...
+        func = self.lib.GEODIFF_createWkbFromGpkgHeader
+        func.argtypes = [ctypes.c_char_p, ctypes.POINTER(ctypes.c_size_t)]
+        func.restype = ctypes.POINTER(ctypes.c_char)
+        free_func = self.lib.GEODIFF_free
+        free_func.argtypes = [ctypes.c_char_p]
+        free_func.restype = None
+
+        size = ctypes.c_size_t(len(geometry))
+        ptr  = func(geometry, ctypes.byref(size))
+        result = ptr[:size.value]
```
Comment (PeterPetrik) [inline pygeodiff/geodifflib.py:448]: does this work, or you need to rather use https://docs.python.org/3/library/copy.html#copy.deepcopy ?

File: pygeodiff/geodifflib.py [hunk 6]
```diff
@@ -435,6 +435,19 @@ def read_changeset(self, changeset):
             raise GeoDiffLibError("Unable to open reader for: " + changeset)
         return ChangesetReader(self, reader_ptr)
 
+    def get_wkb_from_geometry(self, geometry):
+        func = self.lib.GEODIFF_createWkbFromGpkgHeader
+        func.argtypes = [ctypes.c_char_p, ctypes.POINTER(ctypes.c_size_t)]
+        func.restype = ctypes.POINTER(ctypes.c_char)
+        free_func = self.lib.GEODIFF_free
+        free_func.argtypes = [ctypes.c_char_p]
+        free_func.restype = None
```
Comment (PeterPetrik) [inline pygeodiff/geodifflib.py:444]: I would registerr GEODIFF_free function in self. so we can use it on other places (in function register_functions)

**Comment by PeterPetrik** [comment]
can you rebase please from master and also add Geodiff context to the new function in this PR?

**Review by PeterPetrik — CHANGES_REQUESTED** [review]
please rebase

File: geodiff/src/geodiff.cpp [hunk 7]
```diff
@@ -1210,7 +1210,13 @@ bool GEODIFF_CT_columnIsPkey( GEODIFF_ContextH /*contextHandle*/, GEODIFF_Change
 
 const char *GEODIFF_createWkbFromGpkgHeader( GEODIFF_ContextH contextHandle, const char *gpkgWkb, size_t *length )
 {
-  auto result = createWkbFromGpkgHeader( std::string( gpkgWkb, *length ) );
+  const Context *context = static_cast<const Context *>( contextHandle );
+  if ( !context )
```
Comment (PeterPetrik) [inline geodiff/src/geodiff.cpp:1214]: as I write, I think the function should return int (geodiff success or error) and the result should go as 3rd argument before lenght

**Comment by alexbruy** [comment]
Tried to address comments. I have no idea whether this is correct way of doing it or not.

File: geodiff/src/drivers/sqliteutils.cpp [hunk 8]
```diff
@@ -642,6 +642,8 @@ std::string createGpkgHeader( std::string &wkb, const TableColumnInfo &col )
 
 std::string createWkbFromGpkgHeader( const Context *context, const std::string &gpkgWkb )
 {
+  ( void )context;
```
Comment (PeterPetrik) [inline geodiff/src/drivers/sqliteutils.cpp:645]: https://github.com/MerginMaps/geodiff/blob/51b636100e46892f80fe7cb1da0758fc80f023f0/geodiff/src/geodiffutils.hpp#L9

File: geodiff/src/geodiff.h [hunk 9]
```diff
@@ -597,14 +597,10 @@ GEODIFF_EXPORT void GEODIFF_V_destroy(
  * returns geometry as WKB. More details at https://www.geopackage.org/spec/#gpb_format
  * gpkgWkb is a geometry encoded according to GeoPackage spec
  * length is a length of the passed geometry buffer
- * returns buffer containing WKB encoded geometry. Caller should free that buffer by calling GEODIFF_free
+ * wkb is ooutput wkb
```
Comment (PeterPetrik) [inline geodiff/src/geodiff.h:600]: typo: "ooutput"; also write that wkb must be already allocated to size `length (- 1?)` (? or what should be the size ?)

File: geodiff/tests/test_geometry_utils.cpp [hunk 10]
```diff
@@ -26,7 +26,13 @@ TEST( GeometryUtilsTest, test_wkb_from_geometry )
   EXPECT_EQ( entry.table->name, "simple" );
   EXPECT_EQ( entry.oldValues[1].type(), Value::TypeBlob );
   std::string gpkgWkb = entry.oldValues[1].getString();
-  std::string wkb = createWkbFromGpkgHeader( gpkgWkb );
+  const char *c_gpkgWkb = gpkgWkb.c_str();
+  size_t length = gpkgWkb.length();
+  char *c_wkb = new char[ length ];
```
Comment (PeterPetrik) [inline geodiff/tests/test_geometry_utils.cpp:31]: you need to delete c_wkb in the test outhewise you got memory leak

File: pygeodiff/geodifflib.py (hunk 5 above)
Comment (alexbruy) [reply pygeodiff/geodifflib.py]: well, seems test passes, so it should be ok.

File: pygeodiff/geodifflib.py [hunk 11]
```diff
@@ -457,6 +456,16 @@ def read_changeset(self, changeset):
             raise GeoDiffLibError("Unable to open reader for: " + changeset)
         return ChangesetReader(self, reader_ptr)
 
+    def get_wkb_from_geometry(self, geometry):
+        func = self.lib.GEODIFF_createWkbFromGpkgHeader
+        func.argtypes = [ctypes.c_void_p, ctypes.c_char_p, ctypes.POINTER(ctypes.c_size_t)]
```
Comment (PeterPetrik) [inline pygeodiff/geodifflib.py:461]: how this even can work? GEODIFF_createWkbFromGpkgHeader has now 4 arrgs and int restype...

Comment (PeterPetrik) [reply pygeodiff/geodifflib.py:461]: but you want this line to reflect it too, no?

File: pygeodiff/geodifflib.py (hunk 5 above)
Comment (PeterPetrik) [reply pygeodiff/geodifflib.py]: @wonder-sk do you know if deepcopy should be used here?

**Pull request merged**